skrub/_dispatch.py

```diff
@@ -0,0 +1,194 @@
+optional dependencies and may not be installed. Therefore, it may not be
+possible to import the type ``polars.DataFrame`` and use it to register an
+implementation for a standard ``@singledispatch`` function.
+
+Once ``dispatch`` has been applied to a function, it has a ``specialize``
+attribute that can be used to register implementations.
+
+>>> from skrub._dispatch import dispatch
+
+>>> @dispatch
+... def my_function(df, other_arg=1.0):
+...     # here we can either provide a default implementation or:
+...     raise NotImplementedError()
+
+my_function now has a ``specialize`` attribute we can use to register
+implementations:
+
+>>> @my_function.specialize("pandas")
+... def _my_function_pandas(df, other_arg=1.0):
+...     print("calling pandas implementation")
+
+The name ``_my_function_pandas`` is arbitrary, we could use anything such as
+``_`` (unless we want to be able to call the specialized implementation
+directly, eg for testing).
+
+The tentative convention in skrub is to use
+``f"_{generic_function_name.lstrip('_')}_{library_name}"``.
+
+Note that we registered this implementation without needing to import pandas.
+In the same way, we can define an implementation for polars even if it is not
+installed.
+
+>>> @my_function.specialize("polars")
+... def _my_function_polars(df, other_arg=1.0):
+...     print("calling polars implementation")
+
+Now when we call ``my_function`` with a dataframe, the correct implementation is
+chosen based on the type of the first argument.
+
+>>> import pandas as pd
+>>> df = pd.DataFrame(dict(a=[3, 4]))
+>>> my_function(df)
+calling pandas implementation
+
+Inside a specialized implementation, it is safe to import the corresponding
+module or to call methods on the first argument that only exist for this
+specific type.
+
+
+>>> @my_function.specialize("polars")
+... def _my_function_polars(df, other_arg=1.0):
+...     print("calling polars implementation")
+...     # we know df is a polars DataFrame, LazyFrame or Series (see how to
+...     # specialize for only one of those types below)
+...     df.shape             # ok
+...     df.to_pandas()       # ok
+...     import polars as pl  # ok
+
+
+Outside such specialization, skrub code **must not** directly access methods of
+a container, as this would rely on backend-specific behavior:
+
+
+>>> def my_other_function(df):
+...     # not ok: will fail if df is a polars DataFrame
+...     df.values
+...     # not ok: only happens to work by chance because both polars and pandas use
+...     # the same name with the same signature and semantics, but could fail for
+...     # other backends added in the future
+...     df.shape
+...     # ok
+...     import skrub._dataframe as ns
+...     ns.shape(df)
+
+
+We can inspect which implementations were registered in ``my_function.registry``.
+
+It is also possible to register a specialization specifically for dataframes or
+for columns:
+
+>>> @my_function.specialize("pandas", "Column")
```

Comment:
At a first glance, I would prefer to enforce keyword, if you further specialization:

```python
@my_function.specialize(lib="pandas", container="dataframes")
```

Reply:
also, the first one (the lib) is not optional but required


btw don't hesitate to suggest a different syntax if you have ideas eg we could have something like

```
@my_function.polars
def _(df):
   ...

@my_other_function.polars("DataFrame")
def _(df):
    ...

# or

@my_other_function.polars_dataframe
def _(df):
    ...
```

to type even less

Reply:
> What would you think of making the first argument kwarg or positional and the second one kwarg-only?

Actually it was exactly what I thought.